content/posts/2024-11-23-gsoc-windows-plugins.md

```diff
@@ -0,0 +1,34 @@
+
+In order to support DLL builds on MinGW, a [python script](https://github.com/llvm/llvm-project/blob/main/llvm/utils/extract_symbols.py) would scan the object files generated during the build and extract the symbol names from them. In order to remain under the 64K limit, the symbols would be filtered by pattern matching. The final set would then be used to create an import library that the consumer could use. This technique not only potentially over-exported symbols, introduced a secondary source of truth for the code, but also relied on the linker to generate fix up thunks as the compiler could not determine if a symbol originated from a shared library or static library. This would add additional overhead on a function call that went through this import library as it would no longer be a simple indirect call. Such a thunk was also not possible for data symbols such as static fields of classes except for MinGW which uses a custom [runtime fixup](https://lists.llvm.org/pipermail/llvm-dev/2021-June/150866.html).
+
+## What We Did
+Some initial work I did was update the LLVM CMake build system to be able to build a LLVM DLL and use clang-cl's [/Zc:dllexportInlines-](https://blog.llvm.org/2018/11/30-faster-windows-builds-with-clang-cl_14.html). Inline declared class methods by default are not compiled unless used, but when dllexport attribute is applied to a class all its methods are compiled and exported by the compiler even if not used. This option makes dllexport skip compiling and exporting inline methods. Which both avoids emitting them in every translation unit that includes the classes header greatly reducing compile times for DLL builds, and more importantly it almost halves the number of symbols exported for LLVM to 28k and Clang DLL to 20k. The one downside is DLLs built with it can’t be consumed by code being built with MSVC since there will be linker errors from missing exports of the inline methods. There is Microsoft Developer Community [issue](https://developercommunity.visualstudio.com/t/implement-zcdllexportinlines-aka-fvisibility-inlin/374754) to add it to MSVC, but it needs more votes for it to be considered.
```

Comment:
```suggestion
Some initial work I did was update the LLVM CMake build system to be able to build a LLVM DLL and use clang-cl's [/Zc:dllexportInlines-](https://blog.llvm.org/2018/11/30-faster-windows-builds-with-clang-cl_14.html). Inline declared class methods by default are not compiled unless used, but when the `__declspec(dllexport)` attribute is applied to a class all its methods are compiled and exported by the compiler even if not used. This option negates this behaviour, preventing inline methods from being compiled and exported. This avoids emitting these methods in every translation unit including the declaration, greatly reducing compile times for DLL builds. More importantly, it almost halves the number of symbols exported for LLVM to 28k and Clang DLL to 20k. The cost of this improvement is that DLLs built with this option cannot be consumed by code being built with MSVC as that code expects these methods to be available externally. There is a Microsoft Developer Community [issue](https://developercommunity.visualstudio.com/t/implement-zcdllexportinlines-aka-fvisibility-inlin/374754) to add it to MSVC, please consider voting for it so that it may be considered by Microsoft for addition to the MSVC toolchain.
```